durable-storage/src/merkle_layer/node.rs

```diff
@@ -0,0 +1,486 @@
+///      / \                  \
+///     B   C                  C
+/// ```
+///
+/// Assumes the balance factor is -2 and the left node's balance factor is +1.
+fn rotate_right_left(node: &mut Arc<MavlNode>) -> Arc<MavlNode> {
+    let node = Arc::make_mut(node);
+
+    let mut right = node.right_mut().take().expect("right child must exist");
+    let right = Arc::make_mut(&mut right);
+
+    let right_left = right
+        .left_mut()
+        .take()
+        .expect("right's left child must exist");
+    right.balance_factor = -std::cmp::max(right_left.balance_factor, 0);
+
+    let new_left = Arc::new(MavlNode {
+        right: right_left.left.clone(),
+        left: node.left.clone(),
+        key: node.key.clone(),
+        data: node.data.clone(),
+        balance_factor: -std::cmp::min(right_left.balance_factor, 0),
+        ..Default::default()
+    });
+
+    right.left = right_left.clone().right.clone();
+
+    Arc::new(MavlNode {
+        left: Some(new_left),
+        right: Some(Arc::new(right.clone())),
+        key: right_left.key.clone(),
+        data: right_left.data.clone(),
+        balance_factor: 0,
+        ..Default::default()
+    })
+}
+
+/// Set the value of the node with a given key.
+pub(super) fn set(
+    root: &mut Option<Arc<MavlNode>>,
+    key: &Key,
+    data: Vec<u8>,
+) -> (Option<Arc<MavlNode>>, bool) {
+    if let Some(node) = root {
+        #[cfg(debug_assertions)]
+        {
+            // SAFETY: The default recursion limit in Rust is 128
+            // see: <https://doc.rust-lang.org/reference/attributes/limits.html#r-attributes.limits.recursion_limit.syntax>
+            const RECURSION_LIMIT: u32 = 128;
+            // This function recurses once for every node it traverses, meaning that the number
+            // of recursions are equal to or less than the height of the node.
+            debug_assert!(node.height() <= RECURSION_LIMIT);
+            // The lower bound on the number of nodes in a valid AVL tree is:
+            // fibonacci(height + 3) - 1
+            // see: <https://www.cs.cornell.edu/courses/cs2112/2020fa/lectures/avl/>
+            //
+            // with height = 128:
+            // fibonacci(128 + 3) - 1 > 1x10^27
+            //
+            // This would require:
+            //  - more nodes than 64-bit systems can address.
+            //  - more disk space than has ever been produced.
+            //  - inserting 2 billion nodes every second since the dawn of the universe.
+        }
+        match node.key.cmp(key) {
+            // The key already exists and should be updated.
+            Ordering::Equal => {
+                let node = Arc::make_mut(node);
+                node.key = key.clone();
+                node.data = data.clone();
+                (Some(Arc::new(node.clone())), false)
+            }
+            Ordering::Greater => {
+                let node = Arc::make_mut(node);
+                let (new_left, grew) = set(node.left_mut(), key, data);
```

Comment:
The in-place mutation causes the returned `new_left` is always a copy of the `node.left_mut()`. I'm not sure the return value is needed.